streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java

```diff
@@ -1381,6 +1364,313 @@ public WriteBufferManager writeBufferManager() {
+        return this;
+    }
+
+    @Override
+    public boolean enablePipelinedWrite() {
+        return dbOptions.enablePipelinedWrite();
+    }
+
+    @Override
+    public Options setUnorderedWrite(final boolean unorderedWrite) {
+        dbOptions.setUnorderedWrite(unorderedWrite);
+        return this;
+    }
+
+    @Override
+    public boolean unorderedWrite() {
+        return dbOptions.unorderedWrite();
+    }
+
+    @Override
+    public Options setSkipCheckingSstFileSizesOnDbOpen(final boolean skipCheckingSstFileSizesOnDbOpen) {
+        dbOptions.setSkipCheckingSstFileSizesOnDbOpen(skipCheckingSstFileSizesOnDbOpen);
+        return this;
+    }
+
+    @Override
+    public boolean skipCheckingSstFileSizesOnDbOpen() {
+        return dbOptions.skipCheckingSstFileSizesOnDbOpen();
+    }
+
+    @Override
+    public Options setWalFilter(final AbstractWalFilter walFilter) {
+        dbOptions.setWalFilter(walFilter);
+        return this;
+    }
+
+    @Override
+    public WalFilter walFilter() {
+        return dbOptions.walFilter();
+    }
+
+    @Override
+    public Options setAllowIngestBehind(final boolean allowIngestBehind) {
+        dbOptions.setAllowIngestBehind(allowIngestBehind);
+        return this;
+    }
+
+    @Override
+    public boolean allowIngestBehind() {
+        return dbOptions.allowIngestBehind();
+    }
+
+    @Override
+    public Options setPreserveDeletes(final boolean preserveDeletes) {
+        dbOptions.setPreserveDeletes(preserveDeletes);
+        return this;
+    }
+
+    @Override
+    public boolean preserveDeletes() {
+        return dbOptions.preserveDeletes();
+    }
+
+    @Override
+    public Options setTwoWriteQueues(final boolean twoWriteQueues) {
+        dbOptions.setTwoWriteQueues(twoWriteQueues);
+        return this;
+    }
+
+    @Override
+    public boolean twoWriteQueues() {
+        return dbOptions.twoWriteQueues();
+    }
+
+    @Override
+    public Options setManualWalFlush(final boolean manualWalFlush) {
+        dbOptions.setManualWalFlush(manualWalFlush);
+        return this;
+    }
+
+    @Override
+    public boolean manualWalFlush() {
+        return dbOptions.manualWalFlush();
+    }
+
+    @Override
+    public Options setCfPaths(final Collection<DbPath> cfPaths) {
+        columnFamilyOptions.setCfPaths(cfPaths);
+        return this;
+    }
+
+    @Override
+    public List<DbPath> cfPaths() {
+        return columnFamilyOptions.cfPaths();
+    }
+
+    @Override
+    public Options setBottommostCompressionOptions(final CompressionOptions bottommostCompressionOptions) {
+        columnFamilyOptions.setBottommostCompressionOptions(bottommostCompressionOptions);
+        return this;
+    }
+
+    @Override
+    public CompressionOptions bottommostCompressionOptions() {
+        return columnFamilyOptions.bottommostCompressionOptions();
+    }
+
+    @Override
+    public Options setTtl(final long ttl) {
+        columnFamilyOptions.setTtl(ttl);
```

Comment:
Comments in the RocksDB code state:

```
// In Level: Non-bottom-level files older than TTL will go through the
//           compaction process.
// In FIFO: Files older than TTL will be deleted.
```

So FIFO compaction is not a pre-requisite.

Reply:
Testing does not harm, but I would avoid to test too many different combinations because this seems to have the potential to become a maintenance hell.

Reply:
I quickly tested it by setting the compaction style to universal and setting the ttl and could not see any exception.

Reply:
To clarify I was definitely not recommending to test arbitrary combinations or anything beyond just this one case where I thought there may be a conflict. Sounds like they just added a feature for ttl to work with non-FIFO compaction, which will actually make many KV users incredibly happy. We may want to leave a note on that one ticket about a ttl KeyValueStore that has had so many +1's from users

Reply:
From the comment above the TTL option under Level compaction will not delete entries if they exceed the time but it only compacts them. If I remember correctly, the ask was to delete entries if they exceed the time limit.

Reply:
Oh, I guess I didn't read it very carefully 😅  In that case we may want to warn users that this option will not, in fact, provide "real" ttl (unless also using FIFO compaction). I'd imagine users may see this new option pop up and assume, as I did, that it meant deleting files older than ttl, ie the definition of ttl...

Also, not totally sure I understand -- does that mean _only_ files older than the ttl will be compacted, or just that if a file remains uncompacted for longer than the ttl it will be "prematurely" compacted? I assume the latter...

Reply:
> In that case we may want to warn users that this option will not, in fact, provide "real" ttl (unless also using FIFO compaction). I'd imagine users may see this new option pop up and assume, as I did, that it meant deleting files older than ttl, ie the definition of ttl...

If you mean to warn users with a log warning, I am not sure if we should log something because we assume that the expectations of the users are probably wrong.

> Also, not totally sure I understand -- does that mean only files older than the ttl will be compacted, or just that if a file remains uncompacted for longer than the ttl it will be "prematurely" compacted? I assume the latter...

I also do not totally understand, but I also think it is the latter.

Reply:
Feels it is also the latter too.

BTW I think the rocksDB's own javadocs are sufficient enough --- if users do want to set these options, they must be advanced enough to know what exactly rocksDB is doing.. as Streams, we just need to keep it in mind ourselves so if someone asked "why setting TTL does not work in Streams rocksDB" we would refer them to read this first :)

Reply:
Fair enough. FWIW I did not mean to literally log a warning, I guess that word is quite overloaded. I just meant we should be wary of users trying to use it in this way, pretty much exactly what Guozhang said with

> "so if someone asked "why setting TTL does not work in Streams rocksDB" we would refer them to read this first :)"